text/0000-async.md

```diff
@@ -0,0 +1,774 @@
+
+```rust
+fn read_frame<'sock>(socket: &'sock TcpStream)
+    -> impl Async<Output = Result<Frame, io::Error>> + 'sock;
+```
+
+Other async functions can *await* this asynchronous value; see the [companion
+RFC] for full details.
+
+In addition to `async fn` definitions, async values can be built using adapters on
+the `Async` trait, much like with `Iterator`s. The standard library includes a
+number of basic adapters (described in the reference below), while some
+particularly interesting variants are iterating in the crates.io ecosystem
+first.
+
+Ultimately asynchronous computations are executed by *tasks*, which are
+lightweight threads. In particular, an *executor* is able to "spawn" a
+`()`-producing `Async` value as an independent task; these tasks are then
+cooperatively scheduled onto one or more operating system threads. The
+`Executor` trait defines this interface, and the `task` module provides a host
+of related definitions needed when manually implementing `Async` values or
+executors.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## `core::task` module
+
+The fundamental mechanism for asynchronous computation in Rust is *tasks*, which
+are lightweight threads of execution; many tasks can be cooperatively scheduled
+onto a single operating system thread.
+
+To perform this cooperative scheduling we use a technique sometimes referred to
+as a "trampoline". When a task would otherwise need to block waiting for some
+event, instead it schedules itself for later wakeup and *returns* to the
+executor running it, which can then run another task. Subsequent wakeups place
+the task back on the executors queue of ready tasks, much like a thread
+scheduler in an operating system.
+
+Attempting to complete a task (or async value within it) is called *polling*,
+and always yields a `Poll` value back:
+
+```rust
+/// Indicates whether a value is available, or if the current task has been
+/// scheduled for later wake-up instead.
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum Poll<T> {
+    /// Represents that a value is immediately ready.
+    Ready(T),
+
+    /// Represents that a value is not ready yet.
+    ///
+    /// When a function returns `Pending`, the function *must* also
+    /// ensure that the current task is scheduled to be awoken when
+    /// progress can be made.
+    Pending,
+}
+```
+
+When a task returns `Poll::Ready`, the executor knows the task has completed and
+can be dropped.
+
+### Waking up
+
+Each task executor provides its own scheduling facilities, and hence needs to
+customize the way task wakeups are handled. As such, there is a
+`std::task::Wake` trait defining wakeup behavior:
+
+```rust
+/// A way of waking up a specific task.
+///
+/// Any task executor must provide a way of signaling that a task it owns
+/// is ready to be `poll`ed again. Executors do so by providing a wakeup handle
+/// type that implements this trait.
+pub trait Wake: Send + Sync {
```

Comment:
Why does this trait need to be moved into `core`? If `UnsafeWake` exists in core, then would it not be possible to leave this trait in a crate?

Reply:
Yes, this could live out of tree to start with. To be clear, though, I do expect more of the futures crate to move into core over time.